imap/monitor.go

```diff
@@ -147,41 +150,46 @@ func checkForNewEmails(im models.IMAP) {
 				}
 			}
 
-			body := string(append(m.Email.Text, m.Email.HTML...)) // Not sure if we need to check the Text as well as the HTML. Perhaps sometimes Text only emails won't have an HTML component?
-			rid := goPhishRegex.FindString(body)
+			rids, err := checkRIDs(m.Email) // Search email Text, HTML, and each attachment for rid parameters
 
-			if rid != "" {
-				rid = rid[5:]
-				log.Infof("User '%s' reported email with rid %s", m.Email.From, rid)
-				result, err := models.GetResult(rid)
-				if err != nil {
-					log.Error("Error reporting GoPhish email with rid ", rid, ": ", err.Error())
-					reportingFailed = append(reportingFailed, m.UID)
-				} else {
-					err = result.HandleEmailReport(models.EventDetails{})
+			if err != nil {
+				log.Errorf("Error searching email for rids from user '%s': %s", m.Email.From, err.Error())
+			} else {
+				if len(rids) < 1 {
+					// In the future this should be an alert in Gophish
+					log.Infof("User '%s' reported email with subject '%s'. This is not a GoPhish campaign; you should investigate it.", m.Email.From, m.Email.Subject)
+				}
+				for rid := range rids {
+					log.Infof("User '%s' reported email with rid %s", m.Email.From, rid)
+					result, err := models.GetResult(rid)
 					if err != nil {
-						log.Error("Error updating GoPhish email with rid ", rid, ": ", err.Error())
+						log.Error("Error reporting GoPhish email with rid ", rid, ": ", err.Error())
+						reportingFailed = append(reportingFailed, m.SeqNum)
 					} else {
-						if im.DeleteReportedCampaignEmail == true {
-							campaignEmails = append(campaignEmails, m.UID)
+						err = result.HandleEmailReport(models.EventDetails{})
+						if err != nil {
+							log.Error("Error updating GoPhish email with rid ", rid, ": ", err.Error())
+						} else {
+							if im.DeleteReportedCampaignEmail == true {
+								campaignEmails = append(campaignEmails, m.SeqNum)
+							}
 						}
 					}
+
 				}
-			} else {
-				// In the future this should be an alert in Gophish
-				log.Debugf("User '%s' reported email with subject '%s'. This is not a GoPhish campaign; you should investigate it.\n", m.Email.From, m.Email.Subject)
 			}
+
 			// Check if any emails were unable to be reported, so we can mark them as unread
 			if len(reportingFailed) > 0 {
```

Comment:
Right now, this fires inside the `for _, m := range msgs {` loop, but it seems as though that's a bug.

Instead, I think the `for` loop should close before we mark any emails as unread or delete them. Imagine the scenario where I get two emails marked as reported in the same run of the monitor. In that case, the first email will be deleted, then in the next iteration of the loop, Gophish will try to delete both the first and the second emails.

This looks like this:

```
time="2020-06-13T08:34:05-05:00" level=info msg="User 'XXXX' reported email with rid FcHeBrc"
time="2020-06-13T08:34:05-05:00" level=info msg="Deleting 1 campaign emails"
time="2020-06-13T08:34:06-05:00" level=info msg="User 'XXXX' reported email with rid 7eXoql6"
time="2020-06-13T08:34:06-05:00" level=info msg="Deleting 2 campaign emails"
```

Reply:
Oh gosh, sloppy! Fixed.